ethereum/statetransition/src/test/java/tech/pegasys/teku/statetransition/validation/BlockValidatorTest.java

```diff
@@ -210,4 +243,51 @@ void shouldReturnInvalidForBlockThatDoesNotDescendFromFinalizedCheckpoint() {
         blockValidator.validate(blockAndState.getBlock());
     assertThat(result).isCompletedWithValueMatching(InternalValidationResult::isReject);
   }
+
+  @TestTemplate
+  void shouldReturnAcceptOnCorrectExecutionPayloadTimestamp(SpecContext specContext) {
+    onlyMerge(specContext);
+
+    storageSystem = InMemoryStorageSystemBuilder.buildDefault(spec);
+    storageSystem.chainUpdater().initializeGenesisWithPayload(false);
+    recentChainData = storageSystem.recentChainData();
+    blockValidator = new BlockValidator(spec, recentChainData);
+
+    final UInt64 nextSlot = recentChainData.getHeadSlot().plus(ONE);
+    storageSystem.chainUpdater().setCurrentSlot(nextSlot);
+
+    SignedBeaconBlock block = storageSystem.chainBuilder().generateBlockAtSlot(nextSlot).getBlock();
+
+    InternalValidationResult result = blockValidator.validate(block).join();
+    assertTrue(result.isAccept());
+  }
+
+  @TestTemplate
+  void shouldReturnInvalidOnWrongExecutionPayloadTimestamp(SpecContext specContext) {
+    onlyMerge(specContext);
+
+    storageSystem = InMemoryStorageSystemBuilder.buildDefault(spec);
+    storageSystem.chainUpdater().initializeGenesisWithPayload(false);
+    recentChainData = storageSystem.recentChainData();
+    blockValidator = new BlockValidator(spec, recentChainData);
+
+    final UInt64 nextSlot = recentChainData.getHeadSlot().plus(ONE);
+    storageSystem.chainUpdater().setCurrentSlot(nextSlot);
+
+    final SignedBlockAndState signedBlockAndState =
+        storageSystem.chainBuilder().generateBlockAtSlot(nextSlot);
+
+    SignedBeaconBlock block = spy(signedBlockAndState.getBlock());
+    BeaconBlock beaconBlock = spy(block.getMessage());
+    BeaconBlockBody beaconBlockBody = spy(beaconBlock.getBody());
+    doReturn(beaconBlock).when(block).getMessage();
+    doReturn(beaconBlockBody).when(beaconBlock).getBody();
+
+    doReturn(Optional.of(specContext.getDataStructureUtil().randomExecutionPayload()))
+        .when(beaconBlockBody)
+        .getOptionalExecutionPayload();
```

Comment:
Using spies to make this work feels pretty ugly. We generally shouldn't need to mock/spy data structure objects like blocks.
I think we'd ideally use the `generateBlockAtSlot(nextSlot, <BlockOptions>)` method and add the ability to specify a specific `ExecutionPayload` for the created block. Might need to also specify the resulting state root since you won't actually be able to apply the block.

I know it's a bunch more work here but it gives us the test utilities we need to do more tests like this which I think is a pretty big win.

Reply:
I spent some time on that (i implemented the setExecutionPayload) but block processor kicks in even if you are just generating a new block without applying it ant it gets rejected.. so appeared to me that we need a way to skip validations to be able to build "wrong" chains.

Reply:
I updated the test. Have a look if the approach makes sense to you.